### PR TITLE
V2.0/chain networking

### PR DESCRIPTION
**Pull request by valentunn**
* Removed NetworkType from domain and data layer. Generally, it was replaced with chainId/Chain (and ChainAsset/chainAssetId)
* Removed TokenType (replaced with ChainAsset)
* Remoted stakingExposures in-memory cache since multi-chain environment made it harder to manage invalidation. Probably will return it back in the future.
* In staking and crowdloans chain and asset are provided by means of the corresponding implementation of `SingletAssetSharedState` interface
* In staking, we insert balances into DB not for only the current account (which can be determined by metaId), but also for companion (controller -> stash, stash -> controller). In the latter case, we cannot easily find an account since we do not know metaId in advance. For this purpose, I query both meta accounts and chain accounts to find matches with supplied accountId.  
* Fixes issues from #385

### CODE REVIEW

File: core-db/src/main/java/jp/co/soramitsu/core_db/dao/AssetDao.kt [hunk 1]
```diff
@@ -6,38 +6,47 @@ import androidx.room.OnConflictStrategy
...
-import jp.co.soramitsu.core_db.model.TokenLocal
+import jp.co.soramitsu.fearless_utils.runtime.AccountId
 import kotlinx.coroutines.flow.Flow
 
-private const val RETRIEVE_ASSET_SQL = """
-           select * from assets as a inner join tokens as t where a.token = t.type
-            and a.accountAddress = :accountAddress and a.token = :type
+private const val RETRIEVE_ASSET_SQL_META_ID = """
+           select * from assets as a inner join tokens as t ON a.tokenSymbol = t.symbol WHERE
+            a.metaId = :metaId and a.chainId = :chainId AND a.tokenSymbol = :symbol
```
Comment (ERussel) [inline core-db/src/main/java/jp/co/soramitsu/core_db/dao/AssetDao.kt:14]: Looks like now it is more safe to use `assetId`

File: feature-wallet-impl/src/main/java/jp/co/soramitsu/feature_wallet_impl/data/repository/WalletRepositoryImpl.kt [hunk 2]
```diff
@@ -58,75 +58,74 @@ class WalletRepositoryImpl(
...
+            }
+        }
     }
 
-    override fun assetFlow(accountAddress: String, type: Token.Type): Flow<Asset> {
-        return assetCache.observeAsset(accountAddress, mapTokenTypeToTokenTypeLocal(type))
-            .map { mapAssetLocalToAsset(it) }
-    }
+    override suspend fun syncAssetsRates() {
+        val firstChain = chainRegistry.currentChains.first().first()
```
Comment (ERussel) [inline feature-wallet-impl/src/main/java/jp/co/soramitsu/feature_wallet_impl/data/repository/WalletRepositoryImpl.kt:79]: Looks like TODO needed

File: runtime/src/main/java/jp/co/soramitsu/runtime/ext/ChainExt.kt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+val Chain.utilityAsset
+    get() = assets.first { it.id == 0 }
+
+val Chain.genesisHash: String
+    get() = id
+
+fun Chain.addressOf(accountId: ByteArray): String {
+    return if (isEthereumBased) {
+        accountId.toHexString(withPrefix = false)
```
Comment (ERussel) [inline runtime/src/main/java/jp/co/soramitsu/runtime/ext/ChainExt.kt:28]: I think for Ethereum address should be with prefix

File: runtime/src/main/java/jp/co/soramitsu/runtime/multiNetwork/chain/remote/ChainFetcher.kt [hunk 4]
```diff
@@ -5,6 +5,6 @@ import retrofit2.http.GET
 
 interface ChainFetcher {
 
-    @GET("https://raw.githubusercontent.com/soramitsu/fearless-utils/master/chains/chains.json")
+    @GET("https://raw.githubusercontent.com/soramitsu/fearless-utils/master/chains/chains_dev.json")
```
Comment (ERussel) [inline runtime/src/main/java/jp/co/soramitsu/runtime/multiNetwork/chain/remote/ChainFetcher.kt:8]: TODO? :)

**Review by ERussel — APPROVED** [review]

File: feature-account-api/src/main/java/jp/co/soramitsu/feature_account_api/data/extrinsic/ExtrinsicService.kt [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    private val extrinsicBuilderFactory: ExtrinsicBuilderFactory,
+) {
+
+    suspend fun submitExtrinsic(
+        chain: Chain,
+        accountId: ByteArray,
+        formExtrinsic: suspend ExtrinsicBuilder.() -> Unit,
+    ): Result<String> = runCatching {
+        val metaAccount = accountRepository.findMetaAccount(accountId) ?: error("No meta account found accessing ${accountId.toHexString()}")
+        val keypair = secretStoreV2.getKeypairFor(chain, metaAccount)
```
[the remark this answers is not in the document]
Comment (valentunn) [reply feature-account-api/src/main/java/jp/co/soramitsu/feature_account_api/data/extrinsic/ExtrinsicService.kt:31]: Change (metaId, chainId) to (metaId, accountId) in SecretStoreV2

**Pull request merged**